tonic/src/transport/service/tls.rs

```diff
@@ -220,50 +246,28 @@ impl TlsAcceptor {
     }
 
     #[cfg(feature = "rustls")]
-    fn load_rustls_private_key(
-        mut cursor: std::io::Cursor<&[u8]>,
-    ) -> Result<PrivateKey, crate::Error> {
-        // First attempt to load the private key assuming it is PKCS8-encoded
-        if let Ok(mut keys) = pemfile::pkcs8_private_keys(&mut cursor) {
-            if keys.len() > 0 {
-                return Ok(keys.remove(0));
-            }
-        }
-
-        // If it not, try loading the private key as an RSA key
-        cursor.set_position(0);
-        if let Ok(mut keys) = pemfile::rsa_private_keys(&mut cursor) {
-            if keys.len() > 0 {
-                return Ok(keys.remove(0));
-            }
-        }
+    pub(crate) fn new_with_rustls_identity(
+        identity: Identity,
+        client_ca_root: Option<Certificate>,
+    ) -> Result<Self, crate::Error> {
+        let (cert, key) = rustls_keys::load_identity(identity)?;
 
-        // Otherwise we have a Private Key parsing problem
-        Err(Box::new(TlsError::PrivateKeyParseError))
-    }
+        let mut config = match client_ca_root {
+            None => ServerConfig::new(NoClientAuth::new()),
+            Some(cert) => {
+                let mut cert = std::io::Cursor::new(&cert.pem[..]);
 
-    #[cfg(feature = "rustls")]
-    pub(crate) fn new_with_rustls_identity(identity: Identity) -> Result<Self, crate::Error> {
-        let cert = {
-            let mut cert = std::io::Cursor::new(&identity.cert.pem[..]);
-            match pemfile::certs(&mut cert) {
-                Ok(certs) => certs,
-                Err(_) => return Err(Box::new(TlsError::CertificateParseError)),
-            }
-        };
+                let mut client_root_cert_store = tokio_rustls::rustls::RootCertStore::empty();
+                match client_root_cert_store.add_pem_file(&mut cert) {
+                    Err(_) => return Err(Box::new(TlsError::CertificateParseError)),
```

Comment:
do we want to attach the error message here?

Reply:
I'll update this - we likely do want the error message.

Reply:
Only kidding... the error type is `()`.